scripts/dmain.py

```
@@ -0,0 +1,177 @@
+        type=str,
+        help="Comma-separated IP addresses of machines",
+    )
+
+    parser.add_argument(
+        "--ssh_cmd",
+        required=False,
+        type=str,
+        default="ssh -f {addr}",
+        help="SSH command. Useful to utilize a pre-shared key with 'ssh -i mykey.pem -f ubuntu@{addr}'. "
+        "The option `-f` should be used for non-interactive session",
+    )
+
+    parser.add_argument(
+        "--env_activate_path",
+        required=True,
+        type=str,
+        help="Path to the virtual environment's `activate` script. It must be the same across all machines",
+    )
+
+    parser.add_argument(
+        "--allenact_path",
+        required=False,
+        type=str,
+        default="allenact",
+        help="Path to allenact top directory. It must be the same across all machines",
+    )
+
+    return parser
+
+
+def get_args():
+    """Creates the argument parser and parses any input arguments."""
+
+    parser = get_argument_parser()
+    args = parser.parse_args()
+
+    return args
+
+
+def get_raw_args():
+    raw_args = sys.argv[1:]
+    filtered_args = []
+    remove = None
+    for arg in raw_args:
+        if remove is not None:
+            remove = None
+        elif arg in [
+            "--runs_on",
+            "--ssh_cmd",
+            "--env_activate_path",
+            "--allenact_path",
+        ]:
+            remove = arg
+        else:
+            filtered_args.append(arg)
+    return filtered_args
+
+
+def wrap_single(text):
+    return f"'{text}'"
+
+
+def wrap_single_nested(text, quote=r"'\''"):
+    return f"{quote}{text}{quote}"
+
+
+def wrap_double(text):
+    return f'"{text}"'
+
+
+def id_generator(size=4, chars=string.ascii_uppercase + string.digits):
+    return "".join(random.choice(chars) for _ in range(size))
+
+
+# Assume code is deployed in all machines and we can ssh into each of the `runs_on` machines through port 22
+# Assume tool is called from allenact project's root directory
```

Can we throw an error if this is not called from the root?

---

Sure.